language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,1204 @@
+#[cfg(any(test, feature = "instruction_synthesis"))]
+use std::collections::HashMap;
+use std::{collections::VecDeque, convert::TryFrom, marker::PhantomData};
+use vm::{
+    access::ModuleAccess,
+    errors::*,
+    file_format::{Bytecode, FunctionHandleIndex, LocalIndex, StructDefinitionIndex},
+    gas_schedule::{AbstractMemorySize, GasAlgebra, GasCarrier, GasUnits},
+    transaction_metadata::TransactionMetadata,
+    IndexKind,
+};
+use vm_runtime_types::{
+    loaded_data::struct_def::StructDef,
+    native_functions::dispatch::{dispatch_native_function, NativeReturnStatus},
+    value::{Locals, ReferenceValue, Struct, Value},
+};
+
+// Data to resolve basic account and transaction flow functions and structs
+lazy_static! {
+    /// The ModuleId for the Account module
+    pub static ref ACCOUNT_MODULE: ModuleId =
+        { ModuleId::new(account_config::core_code_address(), Identifier::new("LibraAccount").unwrap()) };
+    /// The ModuleId for the Event
+    pub static ref EVENT_MODULE: ModuleId =
+        { ModuleId::new(account_config::core_code_address(), Identifier::new("Event").unwrap()) };
+
+}
+
+// Names for special functions and structs
+lazy_static! {
+    static ref CREATE_ACCOUNT_NAME: Identifier = Identifier::new("make").unwrap();
+    static ref ACCOUNT_STRUCT_NAME: Identifier = Identifier::new("T").unwrap();
+    static ref EMIT_EVENT_NAME: Identifier = Identifier::new("write_to_event_store").unwrap();
+}
+
+/// `Interpreter` instances can execute MOVE functions.
+///
+/// An `Interpreter` instance is a stand alone execution context for a function.
+/// It mimics execution on a single thread, with an call stack and an operand stack.
+/// The `Interpreter` receives a reference to a data store used by certain opcodes
+/// to do operations on data on chain and a `TransactionMetadata` which is also used to resolve
+/// specific opcodes.
+/// A `ModuleCache` is also provided to resolve external references to code.
+// REVIEW: abstract the data store better (maybe a single Trait for both data and event?)
+// The ModuleCache should be a Loader with a proper API.
+// Resolve where GasMeter should live.
+pub struct Interpreter<'alloc, 'txn, P>
+where
+    'alloc: 'txn,
+    P: ModuleCache<'alloc>,
+{
+    /// Operand stack, where MOVE `Value`s are stored for stack operations.
+    operand_stack: Stack,
+    /// The stack of active functions.
+    call_stack: CallStack<'txn>,
+    /// Gas metering to track cost of execution.
+    gas_meter: GasMeter,
+    /// Transaction data to resolve special bytecodes (e.g. GetTxnSequenceNumber, GetTxnPublicKey,
+    /// GetTxnSenderAddress, ...)
+    txn_data: TransactionMetadata,
+    /// List of events "fired" during the course of an execution.
+    // REVIEW: should this live outside the Interpreter?
+    event_data: Vec<ContractEvent>,
+    /// Data store
+    // REVIEW: maybe this and the event should go together as some kind of external context?
+    data_view: TransactionDataCache<'txn>,
```

Comment:
Yeah, I think that makes sense -- could the event data just be part of `TransactionDataCache`?

Reply:
yeah we are on the same page. I am not sure where the right place for all of that is but it seems to make a lot of sense for the global storage to live outside of the VM and being plugged in somewhat. `TransactionDataCache` is an definitely a good option